network/src/protocols/rpc/mod.rs

```diff
@@ -337,6 +322,270 @@ impl InboundRpcs {
+        });
+        let (ack_tx, _) = oneshot::channel();
+        write_reqs_tx.send((message, ack_tx)).await?;
+
+        // Collect counters for requests sent.
+        counters::rpc_messages(network_context, REQUEST_LABEL, SENT_LABEL).inc();
+        counters::rpc_bytes(network_context, REQUEST_LABEL, SENT_LABEL).inc_by(req_len);
+
+        // Create channel over which response is delivered to outbound_rpc_task.
+        let (response_tx, response_rx) = oneshot::channel::<RpcResponse>();
+
+        // Store send-side in the pending map so we can notify outbound_rpc_task
+        // when the rpc response has arrived.
+        self.pending_outbound_rpcs.insert(request_id, response_tx);
+
+        // A future that waits for the rpc response with a timeout. We create the
+        // timeout out here to start the timer as soon as we push onto the queue
+        // (as opposed to whenever it first gets polled on the queue).
+        let mut wait_for_response = tokio::time::timeout(timeout, response_rx).map(|result| {
+            // Flatten errors.
+            match result {
+                Ok(Ok(response)) => Ok(Bytes::from(response.raw_response)),
+                Ok(Err(oneshot::Canceled)) => Err(RpcError::UnexpectedResponseChannelCancel),
+                Err(_elapsed) => Err(RpcError::TimedOut),
+            }
+        });
+
+        // A future that waits for the response and sends it to the application.
+        let notify_application = async move {
+            let mut cancellation = application_response_tx.cancellation().fuse();
+            futures::select! {
+                maybe_response = wait_for_response => {
+                    // TODO(philiphayes): Clean up RpcError. Effectively need to
+                    // clone here to pass the result up to application layer, but
+                    // RpcError is not currently cloneable.
+                    let result_copy = match &maybe_response {
+                        Ok(response) => Ok(response.len() as u64),
+                        Err(err) => Err(RpcError::Error(anyhow!(err.to_string()))),
+                    };
+                    // Notify the application of the results.
+                    application_response_tx.send(maybe_response).map_err(|_| RpcError::UnexpectedResponseChannelCancel)?;
+                    result_copy
+                }
+                _ = cancellation => Err(RpcError::UnexpectedResponseChannelCancel),
+            }
+        };
+
+        let outbound_rpc_task = async move {
+            // Always return the request_id so we can garbage collect the
+            // pending_outbound_rpcs map.
+            match notify_application.await {
+                Ok(response_len) => {
+                    let latency = timer.stop_and_record();
+                    (request_id, Ok((latency, response_len)))
+                }
+                Err(err) => {
+                    // don't record
+                    timer.stop_and_discard();
```

Comment:
Yeah I was having a bit of trouble with this... I think I'll do a pass cleaning up the Peer metrics a bit after this.

Reply:
We should probably do a larger thought process on the codebase as a whole if we want to include errors in latency metrics.

It can be useful, but also bad, where you can track latency and see if it fails quickly, that there's quick ending errors or if it fails at the timeout amount, that the cause is timeouts.  But then it skews the metrics the rest of the time.  I wouldn't worry about it in this refactoring, but maybe we should followup later about it.